### ISSUE TITLE
Random affine transformations?

### ISSUE THREAD

**Issue by hughperkins** [torch/nn#302]
Seems we are missing affine transformations?  eg:
- translate
- rotate
- scale
- and might not technically be an affine transformation, but also cutting a random patch

Note that during learning, these layers would all apply a random translate/rotate/etc, as appropriate ot the layer.  During prediction, typically zero transformatoin would be applied.

**Comment by ajtulloch** [comment]
https://github.com/torch/image#image.simpletrans

**Comment by soumith** [comment]
Do you want the gradients to be backpropagated appropriately as well?
If you want to stick this in the middle of the neural net, it is much more complicated to implement (because the affine transform backprop will have a bunch of race conditions on GPU).

If it is just for the input, the link andrew pointed is sufficient.

**Comment by hughperkins** [comment]
Just for input.  But it should ideally be a module in the network I would think, and apply the transformations randomly during training.  (This is how it works in DeepCL by the way, ie you just put `-rt4` in the network definition to randomly translate up to 4 pixels in each direction https://github.com/hughperkins/DeepCL/blob/master/doc/Commandline.md )

**Comment by mrkulk** [comment]
@soumith and @hughperkins 

I had implemented the backprop (with GPU kernels) for this following Tijmen's thesis (http://www.cs.toronto.edu/~tijmen/tijmen_thesis.pdf)

github repo with code -- https://github.com/mrkulk/inverse-graphics-networks
(see ACR.lua which is defined in Tijmen's thesis and the corresponding test files)

I have not had the time to clean it up but I think it might be useful for you.

**Comment by hughperkins** [comment]
@soumith :
- in deepcl actually the start and end of my network both run in cpu, ie calculating the softmax is in cpu, and hardly takes any time at all; ditto for the translations and patches at the start, which take negligible time compared to the whole convolutional thing
- on the other hand, in torch gpu, seems like we change the entire stack to gpu, no way to somehow make the start or end layers stay on the cpu?
- question: how to detect training or testing from inside a network module?
- in deepcl, I dont backpropagate the gradient unless the previous layer actually asks for it.  This saves quite a lot of execution time actually... how can we do the same thing in a torch net module?

**Comment by soumith** [comment]
@mrkulk looks nice. a cleanup + PR to the nn/cunn repos whenever you have time would be quite awesome.

**Comment by soumith** [comment]
@hughperkins 

* In torch, you can do the same. You call :cuda() only on your main network, and then add it to an nn.Sequential that contains your nn.Jitter layer, followed by an nn.Copy('FloatTensor', 'CudaTensor')
* Same with the end side of things.

* You can detect training/testing by checking for self.train. self.train = true is training mode. https://github.com/torch/nn/blob/master/doc/module.md#training

* A loosely held semantic is that if you nil self.gradInput of a particular layer, the gradient does not get backpropagated back to the input.
Example: https://github.com/torch/nn/blob/master/SpatialConvolution.lua#L106

**Comment by hughperkins** [comment]
For the first three, ok sounds good :-)  I hadnt see nthe jitter layer.  Will hunt for that :-)

> A loosely held semantic is that if you nil self.gradInput of a particular layer, the gradient does not get backpropagated back to the input.  

Presumably we would have to do that ourselves, from outside the network, right?  The modules cant automatically communicate with each other, because they lack any way of obtaining a reference to the next/previous module?

**Comment by soumith** [comment]
There isn't a publicly available Jitter module, I was giving an example. It would be easy to write using image.* functions inside it.

We cant obtain a reference to the next/previous module, because when having graphs where a node that  has two edges going out, there isn't a single "next module". either ways modules are meant to be self-contained logic that is oblivious to the outside world.

**Comment by hughperkins** [comment]
> There isn't a publicly available Jitter module, I was giving an example. It would be easy to write using image.* functions inside it.

Ah, ok.  Yes. And yes, right :-)  So, just to check, if I did write a Jitter module, to what extent would that 'fit in' with the philosophy behind existing modules?  (I just ask, since I created FullyConnected layer, and then realized that it doesnt quite fit in to the way that things work in clnn :-) )

**Comment by hughperkins** [comment]
> We cant obtain a reference to the next/previous module, because when having graphs where a node that has two edges going out, there isn't a single "next module". either ways modules are meant to be self-contained logic that is oblivious to the outside world.

Hmmm, yes.  Makes a certain amount of sense.  One of the appealing things about torch I confess is the simplicity of how things slot together :-)

**Comment by hughperkins** [comment]
Seems like this is really old.  And seems I don't need this currently.

=> closing

**Comment by zhanghang1989** [comment]
Hi, I know this issue is closed. I have implemented a GPU backend package for data augmentation (supporting affine) and I will add more features to it. Please see the [link to HzProc](https://github.com/zhanghang1989/hzproc). An example is as following: 

``` torch
require "hzproc"
require 'image'

-- load the image
I = image.lena():cuda()
image.display(I)

local function hzproc_testAffine()
	-- affine transformation matrix
	mat = torch.Tensor({{1, 0, 0}, {.5, 1, 0}, {0, 0, 1}})
	mat = mat:cuda()
	-- affine mapping
	O = hzproc.AffineMap(I, mat);
	-- display the images
	image.display(O)
end

hzproc_testAffine()
```
